added possibility to freeze layers during training

mmmikael opened the pull request:
Following the discussion in #622, here is a simple strategy to "freeze" specific layers during training. IT works for ```Sequential``` and ```Graph``` models.

It adds a ```trainable``` attribute to the core ```Layer``` class and check its value when creating the model. If a layer's ```trainable``` flag is modified after addition to a model, the model needs to be rebuilt and recompiled. For this, a ```rebuild``` function has been added to ```Sequential``` and ```Graph``` classes.

fchollet commented on keras/layers/core.py:
```diff
@@ -46,6 +48,17 @@ def build(self):
         pass
 
     @property
+    def trainable(self):
```
What is the justification for using a decorator over a boolean attribute `self.trainable` (set by default to True)?

mmmikael replied:
This was to avoid adding a ```trainable``` attribute in the serialization when the boolean is not changed.

fchollet commented:
The `trainable` feature is looking good. But is it unclear what the role of `rebuild` is from a UX perspective. It should probably end with `compile`, and should do more, such as offering an option to reinitialize the weights. Otherwise its role is so tied to the dynamic use of `trainable` as not to make sense in the bigger picture (what would be some other use cases in which one would want to call "rebuild"? What is the *UX* justification for having both `compile` and `rebuild`, as they sound conceptually similar?).

If you want to speed things up, you can submit the `trainable` feature in a separate PR.

jmhessel commented:
Also, a potentially relevant generalization of trainable: perhaps we could have a parameter for each layer containing a multiplicative factor associated with the parameter updates at that layer?

fchollet commented:
It's a cool idea, and it could allow us to skip the `rebuild` function entirely. We could maintain a `param_weights` array of the same size as `params` and apply these weights to `updates` in the compile function (for optimization we could skip an update entirely when the weight is 0). Then only `compile` would be needed after a layer weight update. Actually the update could be entirely dynamical if we used shared_variables as param weights.

Then again... it sounds complicated. Not sure we want to add this extra complexity. 

In any case: let's get rid of `rebuild`, because `compile` can be slightly modified to suffice. 

Let's go the way of greater simplicity : )

jmhessel commented:
That does sound like a good plan for now. A lot of transfer learning works better with decreased (but nonzero) learning rates applied to pretrained layers, so it might be worth keeping in mind... But yeah, there are probably more pressing matters.

mmmikael commented:
Thanks for the feedback.  Yes, I think we can remove ```rebuild``` and actually just modify ```get_params``` in the base class ```Layer```. We don't even need to touch ```compile```. It looks like something like this at the end of ```get_params``` should be enough:
```python
if self.trainable:
   return [], regularizers, consts, []
```

NB: I am not using regularizers or constraints so I am not sure how to deal with those.

mmmikael commented:
Following your comments, I made a few changes: 
* there is now a ```get_params``` function in the ```Sequential``` and ```Graph``` containers that gets called at the beginning of their respective ```compile```.
* no more ```rebuild```

fchollet commented:
Several tests for the Sequential model are now failing. 

> NB: I am not using regularizers or constraints so I am not sure how to deal with those.

If the layer is not trainable, they should be empty too. In particular, `constraints` should always keep the same length as `params` (there is a one to one matching between the two), so returning something while having an empty `params` list would cause issues.

fchollet commented on keras/layers/containers.py:
```diff
@@ -20,11 +20,6 @@ class Sequential(Layer):
 
     def __init__(self, layers=[]):
         self.layers = []
-        self.params = []
```
Why are these attributes being remove? It causes tests to fail, and they are definitely necessary.

fchollet commented:
I'm not sure about the architecture here. It could definitely be useful to have individual access to individual attributes `params`, etc, outside of `get_params`. What are your thoughts?

mmmikael commented:
That was a little bit too radical I guess.

We could add individual access to the four attributes in Layer and add decorators for them in Sequential and Graph where they would be composed dynamically. What do you think?



> Le 16 oct. 2015 à 22:18, François Chollet <notifications@github.com> a écrit :
> 
> I'm not sure about the architecture here. It could definitely be useful to have individual access to individual attributes params, etc, outside of get_params. What are your thoughts?
> 
> —
> Reply to this email directly or view it on GitHub.
>

fchollet commented:
> We could add individual access to the four attributes in Layer and add decorators for them in Sequential and Graph where they would be composed dynamically. What do you think?

That sounds very reasonable.

mmmikael commented:
Now, all tests pass. I made the following changes:
* added individual getters for ```params```, ```regularizers```, ```constraints``` and ```updates``` in ```Layer```
* added decorators for them in ```Sequential``` and ```Graph```

Since there is already a function ```get_params``` that returns all four properties in ```Layer```, the getter for ```params``` is called ```get_param```... (not totally happy with the naming here).

fchollet commented:
The `get_param` method does seem very confusing. Why not cover the `if not trainable` logic in the container's `params` method/property? It would seem to be the simplest thing to do.

Otherwise, if you do want to go in that direction, more work would be needed to make it clean: 
- a `_params` attribute in all layers, and a `set_params` methods to set its value
- a `params` property decorator in all layers, that returns either `[]` or `_params`

But really I would advocate changing the `params` decorator in the containers instead. Much simpler.

mmmikael commented:
> Why not cover the if not trainable logic in the container's params method/property?

We could do that but the reason why I added individual getters to ```Layer``` was to avoid calling ```get_params``` for each method/property in the containers. Moving the check on ```trainable``` to the containers will not change this.

I kept the function ```get_params``` for  backward compatibility but the simplest solution I think is just to get rid of it and have one specific for ```params```. Is there any reason to group the four attributes ```params```, ```regularizers```, ```constraints```, ```updates```?

fchollet commented:
> Is there any reason to group the four attributes params, regularizers, constraints, updates?

Just API simplicity.

> We could do that but the reason why I added individual getters to Layer was to avoid calling get_params for each method/property in the containers. 

The computational overhead is essentially zero, so I wouldn't be worried about this. This is small python-side operation that will generally be called ~1-2 times per run...

The reason I am suggesting this is to reduce as much as possible the amount of changes required to merge this. Always go for maximum simplicity : )

mmmikael commented:
I'm not against simplicity :).

The decorators in the ```Graph``` container would like this:
```python
    @property
    def params(self):
        params = []
        for l in self.nodes.values():
            if l.trainable:
                params += l.get_params()[0]
        return params

    @property
    def regularizers(self):
        regularizers = []
        for l in self.nodes.values():
            if l.trainable:
                regularizers += l.get_params()[1]
        return regularizers
...
```
Is the order of the parameters important for Thenao or just being consistent for all four attributes is enough?

mmmikael commented:
> Is the order of the parameters important for Thenao or just being consistent for all four attributes is enough?

I made the node dictionary an ```OrderedDict``` in case it matters.

fchollet commented:
Great, looks good to me! Merging.

EderSantana commented:
Does this require any change on layers definition?

mmmikael commented:
> Does this require any change on layers definition?

No, that should not affect default layer definition.

EderSantana commented:
awesome! tkx!

dbonadiman commented:
Hello great work there i missed this pull request :D 

However i want to ask if there is any possibility to add the flag trainable in `add` function for `Sequential`containers and in the `add_node` for the `Graph` model. From the UX prospective i mean.

mmmikael commented:
> However i want to ask if there is any possibility to add the flag trainable in add function for Sequentialcontainers and in the add_node for the Graph model. From the UX prospective i mean.

This seems quite straightforward to implement. @fchollet, @EderSantana what do you think?

It could be a tri-state ```trainable``` argument: ```None``` (default), ```True```, ```False```. ```None``` would be useful in cases where the layer's trainable boolean is being set beforehand.

PiranjaF commented:
I have a Merge layer combining 2 paths in my network. The network is loaded from a previously trained network (saved as .json and .h5). I'd like to make the layers in one of these paths untrainable, but if I try I get I get the error `Merge object does not support indexing`. Can this be fixed somehow?

mmmikael commented:
@PiranjaF can you open an issue with more details on the error?
